modules/assim.sequential/R/SDA_OBS_Assembler.R

```diff
@@ -0,0 +1,175 @@
+          Ind_list_match <- c(Ind_list_match, j)
+        }
+      }
+    }
+    
+    #fill in more args in the Obs_Prep object.
+    Temp_unlist <- unlist(Obs_Prep)
+    for (j in 1:length(fun_args)) {
+      if(j %in% Ind_list_match) next #if we already assigned then jump to the next.
+      Ind_single_match <- grep(stringr::str_replace_all(fun_args[j], "[^[:alnum:]]", ""), 
+                               stringr::str_replace_all(names(Temp_unlist), "[^[:alnum:]]", ""),
+                                  ignore.case = T)
+      if(length(Ind_single_match)>1){
+        Ind_single_match <- grep(stringr::str_replace_all(paste0(var, fun_args[j]), "[^[:alnum:]]", ""), 
+                                 stringr::str_replace_all(names(Temp_unlist), "[^[:alnum:]]", ""),
+                                 ignore.case = T)
+      }
+      args[[fun_args[j]]] <- as.character(Temp_unlist[Ind_single_match])
+    }
+    #function calls
+    OBS[[i]] <- do.call(obs_prep_fun, args)[[1]]
+    new_var <- c(new_var, var)
+  }
+  
+  #Create obs.mean and obs.cov
+  obs.mean <- obs.cov <- list()
+  new_diag <- function(vec){
+    if(length(vec)==1){
+      return(vec)
+    }else{
+      return(diag(vec))
+    }
+  }
+  rm_na_diag <- function(mat){
+    mat[-which(is.na(rowSums(mat))), -which(is.na(colSums(mat)))]
+  }
+  
+  #over time
+  for (i in 1:length(time_points)) {
+    t <- time_points[i]
+    dat_all_var <- sd_all_var <- matrix(NA, length(Site_Info$site_id), length(new_var)) %>% `colnames<-`(new_var)
+    #over variable
+    for (j in 1:length(OBS)) {
+      if(paste0(t, "_", Var[j]) %in% colnames(OBS[[j]])){
+        dat_all_var[,j] <- OBS[[j]][,paste0(t, "_", Var[j])]
+        sd_all_var[,j] <- OBS[[j]][,paste0(t, "_SD")]^2 #convert from SD to Var
+      }else{
+        dat_all_var[,j] <- NA
+        sd_all_var[,j] <- NA
+      }
+    }
+    #over site
+    site_dat_var <- site_sd_var <- list()
+    for (j in 1:dim(dat_all_var)[1]) {
+      site_dat_var[[j]] <- dat_all_var[j,] %>% matrix(1,length(new_var)) %>% data.frame %>% `colnames<-`(new_var)
+      site_sd_var[[j]] <- new_diag(sd_all_var[j,])
+    }
+    obs.mean[[i]] <- site_dat_var %>% purrr::set_names(Site_Info$site_id)
+    obs.cov[[i]] <- site_sd_var %>% purrr::set_names(Site_Info$site_id)
+  }
+  names(obs.mean) <- gsub("-", "/", time_points)#not sure if I have to do this.
+  names(obs.cov) <- gsub("-", "/", time_points)
+  
+  #remove NA data as this will crash the SDA. Removes rown numbers (may not be nessesary)
+  for (i in 1:length(obs.mean)) {
+    for (j in 1:length(obs.mean[[i]])) {
+      if(sum(is.na(obs.mean[[i]][[j]]))){
+        obs.mean[[i]][[j]] <- obs.mean[[i]][[j]][-which(is.na(obs.mean[[i]][[j]]))]
+        obs.cov[[i]][[j]] <- obs.cov[[i]][[j]][-which(is.na(rowSums(obs.cov[[i]][[j]]))), -which(is.na(colSums(obs.cov[[i]][[j]])))]
+      }
+    }
+  }
+  
+  #rename variables to the var/names in settings.
+  for (i in 1:length(time_points)) {
```

Comment:
1. Not sure why this is needed. Why not just call things what they should be from the start rather than doing this sort of replacement?
2. This hardcoding of cases isn't scalable.

Reply:
Fixed.